tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py

```diff
@@ -395,3 +397,183 @@ def test_invalid(self):
+        class BaseMatlike(_interface.LinearOperator):
+            args = ()
+
+            def __init__(self, dtype):
+                self.dtype = cupy.dtype(dtype)
+                self.shape = original.shape
+
+            def _matvec(self, x):
+                return mv(x, self.dtype)
+
+        # using above base-class to make and test sub-classes based
+        # on the current `inner_modification` parameter
+        if(self.inner_modification == 'adjoint'):
+
+            # defining the sub-class
+            class HasAdjoint(BaseMatlike):
+                args = ()
+
+                def _adjoint(self):
+                    shape = self.shape[1], self.shape[0]
+                    matvec = partial(rmv, dtype=self.dtype)
+                    rmatvec = partial(mv, dtype=self.dtype)
+                    return _interface.LinearOperator(matvec=matvec,
+                                                     rmatvec=rmatvec,
+                                                     dtype=self.dtype,
+                                                     shape=shape)
+            return (HasAdjoint(dtype), original)
+
+        # defining the sub-classes
+        class HasRmatvec(BaseMatlike):
+            args = ()
+
+            def _rmatvec(self, x):
+                return rmv(x, self.dtype)
+
+        class HasRmatmat(HasRmatvec):
+
+            def _matmat(self, x):
+                return original.dot(x)
+
+            def _rmatmat(self, x):
+                return original.T.conj().dot(x)
+
+        if(self.inner_modification == 'adjoint-matrix-vector'):
+            return (HasRmatvec(dtype), original)
+        if(self.inner_modification == 'adjoint-matrix-matrix'):
+            return (HasRmatmat(dtype), original)
+
+    def _outer_cases(self):
+        if(self.dtype != cupy.complex_):
+            original = cupy.array([[1., 2., 3.], [4., 5., 6.]])
+        else:
+            original = cupy.array([[1, 2j, 3j], [4j, 5j, 6]])
+
+        if(self.outer_modification == 'normal'):
+            return self._inner_cases(original, self.dtype)
+        if(self.outer_modification == 'transpose'):
+            M, A = self._inner_cases(original.T, self.dtype)
+            return (_interface.aslinearoperator(M).T, A.T)
+        if(self.outer_modification == 'hermitian'):
+            M, A = self._inner_cases(original.T, self.dtype)
+            return (_interface.aslinearoperator(M).H, A.T.conj())
+
+    def test_basic(self):
+        M, A_array = self._outer_cases()
+        A = _interface.aslinearoperator(M)
+        M, N = A.shape
+
+        xs = [cupy.array([1, 2, 3]),
+              cupy.array([[1], [2], [3]])]
+        ys = [cupy.array([1, 2]), cupy.array([[1], [2]])]
+
+        if A.dtype == cupy.complex_:
+            xs += [cupy.array([1, 2j, 3j]),
+                   cupy.array([[1], [2j], [3j]])]
+            ys += [cupy.array([1, 2j]), cupy.array([[1], [2j]])]
+
+        x2 = cupy.array([[1, 4], [2, 5], [3, 6]])
+
+        for x in xs:
```

Comment:
Please split `test_basic` to `test_matvec`, `test_matmat`... and so on.

Reply:
gotcha